controllers/machine_controller_node.go

```diff
@@ -0,0 +1,161 @@
+	}
+
+	// Set the Machine NodeRef.
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &corev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
+			Name:       node.Name,
+			UID:        node.UID,
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, corev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+	}
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	status, message := summarizeNodeConditions(node)
+	if status == corev1.ConditionFalse {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, message)
+		return reconcile.Result{}, nil
+	}
+
+	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
+	return reconcile.Result{}, nil
+}
+
+// summarizeNodeConditions summarizes a Node's conditions and returns the summary of condition statuses and concatenate failed condition messages:
+// if there is at least 1 semantically-negative condition, summarized status = False;
+// if there is at least 1 semantically-positive condition when there is 0 semantically negative condition, summarized status = True;
+// if all conditions are unknown,  summarized status = Unknown.
+// (semantically true conditions: NodeMemoryPressure/NodeDiskPressure/NodePIDPressure == false or Ready == true.)
+func summarizeNodeConditions(node *corev1.Node) (corev1.ConditionStatus, string) {
+	totalNumOfConditionsChecked := 4
+	semanticallyFalseStatus := 0
+	unknownStatus := 0
+
+	message := ""
+	for _, condition := range node.Status.Conditions {
+		switch condition.Type {
+		case corev1.NodeMemoryPressure, corev1.NodeDiskPressure, corev1.NodePIDPressure:
+			if condition.Status != corev1.ConditionFalse {
+				message += fmt.Sprintf("Condition %s is %s", condition.Type, condition.Status) + ". "
+				if condition.Status == corev1.ConditionUnknown {
+					unknownStatus++
+					continue
+				}
+				semanticallyFalseStatus++
+			}
+		case corev1.NodeReady:
+			if condition.Status != corev1.ConditionTrue {
+				message += fmt.Sprintf("Condition %s is %s", condition.Type, condition.Status) + ". "
+				if condition.Status == corev1.ConditionUnknown {
+					unknownStatus++
+					continue
+				}
+				semanticallyFalseStatus++
+			}
+		}
+	}
+	if semanticallyFalseStatus > 0 {
+		return corev1.ConditionFalse, message
+	}
+	if semanticallyFalseStatus+unknownStatus < totalNumOfConditionsChecked {
+		return corev1.ConditionTrue, message
+	}
+	return corev1.ConditionUnknown, message
+}
+
+func (r *MachineReconciler) getNode(c client.Reader, providerID *noderefutil.ProviderID) (*corev1.Node, error) {
+	logger := r.Log.WithValues("providerID", providerID)
+
+	nodeList := corev1.NodeList{}
+	for {
+		if err := c.List(context.TODO(), &nodeList, client.Continue(nodeList.Continue)); err != nil {
```

Comment:
This should be a note for a future PR, but I think that searching node for `node.Spec.ProviderID` is a common pattern in CAPI, so might be worth to create an index on it...